File: library/src/main/java/org/xmtp/android/library/codecs/GroupUpdatedCodec.kt
package org.xmtp.android.library.codecs

import org.xmtp.proto.mls.message.contents.TranscriptMessages.GroupUpdated

typealias GroupUpdated = org.xmtp.proto.mls.message.contents.TranscriptMessages.GroupUpdated

val ContentTypeGroupUpdated = ContentTypeIdBuilder.builderFromAuthorityId(
    "xmtp.org",
    "group_updated",
    versionMajor = 1,
    versionMinor = 0,
)

data class GroupUpdatedCodec(override var contentType: ContentTypeId = ContentTypeGroupUpdated) :
    ContentCodec<GroupUpdated> {

    override fun encode(content: GroupUpdated): EncodedContent {
        return EncodedContent.newBuilder().also {
            it.type = ContentTypeGroupUpdated
            it.content = content.toByteString()
        }.build()
    }

    override fun decode(content: EncodedContent): GroupUpdated {
        return GroupUpdated.parseFrom(content.content)
    }

    override fun fallback(content: GroupUpdated): String? {
        return null
    }

    override fun shouldPush(content: GroupUpdated): Boolean = false
}

File: sdks/ios/Sources/XMTPiOS/Codecs/GroupUpdatedCodec.swift
//
//  GroupUpdatedCodec.swift
//
//
//  Created by Pat Nakajima on 2/1/24.
//

import Foundation

public typealias GroupUpdated = Xmtp_Mls_MessageContents_GroupUpdated

public let ContentTypeGroupUpdated = ContentTypeID(
	authorityID: "xmtp.org",
	typeID: "group_updated",
	versionMajor: 1,
	versionMinor: 0,
)

public struct GroupUpdatedCodec: ContentCodec {
	public typealias T = GroupUpdated

	public init() {}

	public var contentType = ContentTypeGroupUpdated

	public func encode(content: GroupUpdated) throws -> EncodedContent {
		var encodedContent = EncodedContent()

		encodedContent.type = ContentTypeGroupUpdated
		encodedContent.content = try content.serializedBytes()

		return encodedContent
	}

	public func decode(content: EncodedContent) throws -> GroupUpdated {
		try GroupUpdated(serializedData: content.content)
	}

	public func fallback(content _: GroupUpdated) throws -> String? {
		nil
	}

	public func shouldPush(content _: GroupUpdated) throws -> Bool {
		false
	}
}
